### DISCUSSION TITLE
실행 컨텍스트에 대해 설명해 주세요

### DISCUSSION THREAD

**Discussion by greenblues1190** [woowacourse-study/2022-woowahan-bansanghwe#89] [JavaScript]
# 실행 컨텍스트

## 실행 컨텍스트(Execution Context)란?

실행 컨텍스트의 정의는 [ECMA-262 문서](https://tc39.es/ecma262/#sec-execution-contexts)에 의하면 다음과 같다.

> An execution context is a specification device that is used to track the runtime evaluation of code by an ECMAScript implementation.
>
> *실행 컨텍스트는 코드의 ECMAScript 구현에 따른 런타임 평가를 추적하기 위해 사용되는 명세 장치입니다. (틀릴 수 있음)*

다시 말해, 실행 컨텍스트는 자바스크립트 엔진이 **런타임**에 소스코드를 평가하고 실행하기 위해 필요한 환경을 제공하고 코드의 실행 결과를 실제로 관리하는 영역이다. 자바스크립트 엔진은 실행 컨텍스트를 통해 식별자와 스코프를 관리하며 런타임에서 변수와 함수의 평가 결과를 동적으로 결정한다.

### 생성 단계

- 실행 컨텍스트를 생성한다.
- 선언문을 평가해 스코프(렉시컬 환경의 환경 레코드)에 기록한다.
  - `var` 키워드로 변수를 선언하는 경우, 변수의 식별자를 환경 레코드에 등록하고 값을 `undefined`로 초기화한다.
  - `let`과 `const`로 변수를 선언하는 경우, 자바스크립트 엔진이 식별자는 등록하되 값을 초기화하지 않는다.

#### 환경 레코드(Environment Records)

> *Environment Record* is a specification type used to define the association of [Identifier](https://tc39.es/ecma262/#prod-Identifier)s to specific variables and functions, based upon the lexical nesting structure of ECMAScript code. Usually an Environment Record is associated with some specific syntactic structure of ECMAScript code such as a [FunctionDeclaration](https://tc39.es/ecma262/#prod-FunctionDeclaration), a [BlockStatement](https://tc39.es/ecma262/#prod-BlockStatement), or a [Catch](https://tc39.es/ecma262/#prod-Catch) clause of a [TryStatement](https://tc39.es/ecma262/#prod-TryStatement). Each time such code is evaluated, a new Environment Record is created to record the identifier bindings that are created by that code.

Environment Record는 다음 3개의 하위 클래스가 있는 추상 클래스이다.

- **Declarative Environment Record**: A Declarative Environment Record is used to define the effect of ECMAScript language syntactic elements such as [FunctionDeclaration](https://tc39.es/ecma262/#prod-FunctionDeclaration)s, [VariableDeclaration](https://tc39.es/ecma262/#prod-VariableDeclaration)s, and [Catch](https://tc39.es/ecma262/#prod-Catch) clauses that directly associate identifier bindings with [ECMAScript language values](https://tc39.es/ecma262/#sec-ecmascript-language-types). 

  Each Declarative Environment Record is associated with an ECMAScript program scope containing variable, constant, let, class, module, import, and/or function declarations. A Declarative Environment Record binds the set of identifiers defined by the declarations contained within its scope.

  - **Function Environment Records**: A Function Environment Records corresponds to the invocation of an ECMAScript [function object](https://tc39.es/ecma262/#function-object), and contains bindings for the top-level declarations within that function. It may establish a new this binding. It also captures the state necessary to support super method invocations.

    A Function Environment Record is a [Declarative Environment Record](https://tc39.es/ecma262/#sec-declarative-environment-records) that is used to represent the top-level scope of a function and, if the function is not an [ArrowFunction](https://tc39.es/ecma262/#prod-ArrowFunction), provides a this binding. If a function is not an [ArrowFunction](https://tc39.es/ecma262/#prod-ArrowFunction) function and references super, its Function Environment Record also contains the state that is used to perform super method invocations from within the function.

  - **Module Environment Records**: A Module Environment Record contains the bindings for the top-level declarations of a [Module](https://tc39.es/ecma262/#prod-Module). It also contains the bindings that are explicitly imported by the [Module](https://tc39.es/ecma262/#prod-Module). Its [[OuterEnv]] is a [Global Environment Record](https://tc39.es/ecma262/#sec-global-environment-records).

    A Module Environment Record is a [Declarative Environment Record](https://tc39.es/ecma262/#sec-declarative-environment-records) that is used to represent the outer scope of an ECMAScript [Module](https://tc39.es/ecma262/#prod-Module). In additional to normal mutable and immutable bindings, Module Environment Records also provide immutable import bindings which are bindings that provide indirect access to a target binding that exists in another [Environment Record](https://tc39.es/ecma262/#sec-environment-records).

- **Object Environment Record**: An Object Environment Record is used to define the effect of ECMAScript elements such as [WithStatement](https://tc39.es/ecma262/#prod-WithStatement) that associate identifier bindings with the properties of some object. Each Object Environment Record is associated with an object called its binding object.
 
  An Object Environment Record binds the set of string identifier names that directly correspond to the property names of its binding object. [Property keys](https://tc39.es/ecma262/#sec-object-type) that are not strings in the form of an [IdentifierName](https://tc39.es/ecma262/#prod-IdentifierName) are not included in the set of bound identifiers. Both own and inherited properties are included in the set regardless of the setting of their [[Enumerable]] attribute. Because properties can be dynamically added and deleted from objects, the set of identifiers bound by an Object Environment Record may potentially change as a side-effect of any operation that adds or deletes properties. Any bindings that are created as a result of such a side-effect are considered to be a mutable binding even if the Writable attribute of the corresponding property is false. Immutable bindings do not exist for Object Environment Records.

  Object Environment Records created for `with` statements ([14.11](https://tc39.es/ecma262/#sec-with-statement)) can provide their binding object as an implicit this value for use in function calls. The capability is controlled by a Boolean `[[IsWithEnvironment]]` field.

- **Global Environment Record**: A Global Environment Record is used for [Script](https://tc39.es/ecma262/#prod-Script) global declarations. It does not have an outer environment; its [[OuterEnv]] is null. It may be prepopulated with identifier bindings and it includes an associated [global object](https://tc39.es/ecma262/#sec-global-object) whose properties provide some of the global environment's identifier bindings. As ECMAScript code is executed, additional properties may be added to the [global object](https://tc39.es/ecma262/#sec-global-object) and the initial properties may be modified.

  A Global Environment Record is used to represent the outer most scope that is shared by all of the ECMA[Script](https://tc39.es/ecma262/#prod-Script) Script elements that are processed in a common [realm](https://tc39.es/ecma262/#realm). A Global Environment Record provides the bindings for built-in globals (clause [19](https://tc39.es/ecma262/#sec-global-object)), properties of the [global object](https://tc39.es/ecma262/#sec-global-object), and for all top-level declarations ([8.2.9](https://tc39.es/ecma262/#sec-static-semantics-toplevellexicallyscopeddeclarations), [8.2.11](https://tc39.es/ecma262/#sec-static-semantics-toplevelvarscopeddeclarations)) that occur within a [Script](https://tc39.es/ecma262/#prod-Script).

  A Global Environment Record is logically a single record but it is specified as a composite encapsulating an [Object Environment Record](https://tc39.es/ecma262/#sec-object-environment-records) and a [Declarative Environment Record](https://tc39.es/ecma262/#sec-declarative-environment-records). The [Object Environment Record](https://tc39.es/ecma262/#sec-object-environment-records) has as its base object the [global object](https://tc39.es/ecma262/#sec-global-object) of the associated [Realm Record](https://tc39.es/ecma262/#realm-record). This [global object](https://tc39.es/ecma262/#sec-global-object) is the value returned by the Global Environment Record's GetThisBinding concrete method. The [Object Environment Record](https://tc39.es/ecma262/#sec-object-environment-records) component of a Global Environment Record contains the bindings for all built-in globals (clause [19](https://tc39.es/ecma262/#sec-global-object)) and all bindings introduced by a [FunctionDeclaration](https://tc39.es/ecma262/#prod-FunctionDeclaration), [GeneratorDeclaration](https://tc39.es/ecma262/#prod-GeneratorDeclaration), [AsyncFunctionDeclaration](https://tc39.es/ecma262/#prod-AsyncFunctionDeclaration), [AsyncGeneratorDeclaration](https://tc39.es/ecma262/#prod-AsyncGeneratorDeclaration), or [VariableStatement](https://tc39.es/ecma262/#prod-VariableStatement) contained in global code. The bindings for all other ECMAScript declarations in global code are contained in the [Declarative Environment Record](https://tc39.es/ecma262/#sec-declarative-environment-records) component of the Global Environment Record.

  Properties may be created directly on a [global object](https://tc39.es/ecma262/#sec-global-object). Hence, the [Object Environment Record](https://tc39.es/ecma262/#sec-object-environment-records) component of a Global Environment Record may contain both bindings created explicitly by [FunctionDeclaration](https://tc39.es/ecma262/#prod-FunctionDeclaration), [GeneratorDeclaration](https://tc39.es/ecma262/#prod-GeneratorDeclaration), [AsyncFunctionDeclaration](https://tc39.es/ecma262/#prod-AsyncFunctionDeclaration), [AsyncGeneratorDeclaration](https://tc39.es/ecma262/#prod-AsyncGeneratorDeclaration), or [VariableDeclaration](https://tc39.es/ecma262/#prod-VariableDeclaration) declarations and bindings created implicitly as properties of the [global object](https://tc39.es/ecma262/#sec-global-object). In order to identify which bindings were explicitly created using declarations, a Global Environment Record maintains a list of the names bound using its CreateGlobalVarBinding and CreateGlobalFunctionBinding concrete methods.

### 실행 단계

- 선언문 외 나머지 코드를 순차적으로 실행한다.
- 스코프를 참조하고, 코드의 실행 과정에서의 변수 값의 변경 등을 실행 컨텍스트가 관리하는 스코프에 업데이트한다.

#### `var` 키워드로 변수를 선언하는 경우

식별자의 값을 `undefined`로 초기화되었기 때문에 선언 라인 전에 참조하면 `undefined`로 평가된다.

#### `let`과 `const`로 변수를 선언하는 경우

값이 초기화되지 않았기 때문에 선언 라인 이전에 참조하면 `ReferenceError`가 발생한다. 선언 라인 이전에 식별자를 참조할 수 없는 구간을 *Temporary Dead Zone*이라 한다. 

#### 함수 표현식으로 함수를 선언하는 경우

함수를 변수에 할당하기 때문에 변수 호이스팅과 같이 동작한다. `var`로 선언하면 `undefined`로 초기화 되어 선언 라인 이전에 실행하는 경우 `TypeError`가 발생한다. `let`과 `const`의 경우 기록된 값이 없기 때문에 `ReferenceError`가 발생한다.

#### 함수 선언문으로 함수를 선언하는 경우

자바스크립트 엔진이 함수의 선언과 동시에 함수 객체를 생성해 환경 레코드에 기록한다. 선언과 동시에 함수가 생성되기 때문에 선언 전에도 함수를 실행할 수 있다.

## `this` 키워드와 실행 컨텍스트

> As discussed previously, when a function is defined, it is attached to its enclosing scope via closure. Scope is the set of rules that controls how references to variables are resolved.
>
> But functions also have another characteristic besides their scope that influences what they can access. This characteristic is best described as an **execution context**, and it's exposed to the function via its this keyword.
>
> Scope is static and contains a fixed set of variables available at the moment and location you define a function, but a function's execution context is dynamic, entirely dependent on how it is called (regardless of where it is defined or even called from).

## 참고

- 모던 자바스크립트 Deep Dive
- [You Don't Know JS Yet: Get Started - 2nd Edition](https://github.com/getify/You-Dont-Know-JS/blob/8861041133f496edce0d03885e2e998d50c3414a/get-started/ch3.md)
- [ECMA-262](https://tc39.es/ecma262/#sec-execution-contexts)
- [[10분 테코톡] 💙 하루의 실행 컨텍스트](https://www.youtube.com/watch?v=EWfujNzSUmw)

**Comment by al-bur** [comment]
실행 컨텍스트에 대해서 설명해주세요

**Comment by al-bur** [comment]
한 두문장으로 실행 컨텍스트를 정의해본다면?

**Comment by al-bur** [comment]
자바스크립트는 왜 실행컨텍스트라는 개념을 도입했을까요?

**Comment by al-bur** [comment]
실행 컨텍스트는 언제 생성될까요?

**Comment by al-bur** [comment]
실행 컨텍스트와 스코프 체인의 연관성

**Comment by al-bur** [comment]
실행 컨텍스트 객체의 구성요소

**Comment by JUDONGHYEOK** [comment]
호이스팅이란?

**Reply by al-bur** [reply]
선언문 이전에 해당 변수가 참조가 되는 현상?

**Comment by al-bur** [comment]
let과 const는 호이스팅이 되지 않는가?

**Comment by al-bur** [comment]
초기화란 무엇일까요?

**Comment by al-bur** [comment]
할당은 어떤 동작인지?

**Comment by al-bur** [comment]
할당 전에는 변수가 메모리 값을 가리키고있나요?

**Comment by al-bur** [comment]
클로저를 실행 컨텍스트 관점에서 설명해주세요

**Comment by al-bur** [comment]
호이스팅의 장점은?

**Comment by al-bur** [comment]
비동기 함수의 실행 컨텍스트 과정? (질문이 안좋다)

**Comment by al-bur** [comment]
this란?

**Comment by al-bur** [comment]
키워드란?

**Comment by al-bur** [comment]
this 동적으로 결정되는 방식은?
